### PR TITLE
[JENKINS-51716] _links could not be defined on some BO objects

### PR DESCRIPTION
**Pull request by kuisathaverat**
# Description

See [JENKINS-51716](https://issues.jenkins-ci.org/browse/JENKINS-51716).

# Submitter checklist
- [X] Link to JIRA ticket in description, if appropriate.
- [X] Change is code complete and matches issue description
- [x] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [X] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
  * Go to the activity page
  * Open dev tool on chrome
  * Set a breakpoint on Pager.fetchNextPage method
  * Reload the page
  * Change the value of the data variable with the data.json file provided in the Jira
  * Continue the execution, now the page is loaded. 

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

**Comment by halkeye** [comment]
@kuisathaverat can you rebase this branch?

Also, can you add a test for your change? see cat test/js/services/LocationService-spec.js for an example. The json file you have in the ticket would be a great test case. Don't worry about the unrelated changes, just the one you made.

If your not comfortable, I'll see if i can help out.

**Comment by kuisathaverat** [comment]
@halkeye I'm going to try to make it today.

**Comment by halkeye** [comment]
Ack, Finally back to reviewing this, we've been busy trying to stabalize our tests and slowly working through older PRs

File: blueocean-core-js/src/js/services/BunkerService.js [hunk 1]
```diff
@@ -33,7 +33,11 @@ export class BunkerService {
      * @returns {any} The key for the store.
      */
     bunkerKey(data) {
+      if(data && data._links && data._links.self && data._links.self.href){
         return data._links.self.href;
+      } else {
+        return Math.random().toString(36).substring(2) + (new Date()).getTime().toString(36);
```
Comment (halkeye) [inline blueocean-core-js/src/js/services/BunkerService.js]: So asking around, all items in the bunker should have a deterministic key. Using a random string fixes the symptom, but doesn't fix the problem.

In your BunkerService-data, you have `"_class": "io.jenkins.blueocean.rest.impl.pipeline.PipelineRunImpl",` for both cases. From your testing (cause i don't really know how to reproduce it), is it only pipelinerunimpl that don't have a self link?

**Comment by imeredith** [comment]
If the Rest API is not returning the self href, the backend needs to be fixed. 

The other potential source is from the SSE events. We sometimes have to generate our own hrefs https://github.com/jenkinsci/blueocean-plugin/blob/master/blueocean-core-js/src/js/sse/SseBus.js

**Comment by kuisathaverat** [comment]
I agree it is a symptom of some issue in the backend, this PR is a dirty workaround, the fix should be made on the backend, and it is probably made because I did not see the issue anymore since a couple of versions of so.

**Pull request closed without merging**